specs/withdrawals.md

```diff
@@ -0,0 +1,176 @@
+  storage root of the Withdrawals predeploy contract at `0x4200000000000000000000000000000000000015` (described below).
+
+1. The `WithdrawalsRootInclusionProof` proof contains the data necessary to prove that the provided `WithdrawalsRoot` is
+  included in the [SSZ merkleization](https://github.com/ethereum/consensus-specs/blob/dev/ssz/simple-serialize.md#merkleization)
+  of the `L2Output` as defined in the
+  [L2 output commitment construction](./proposals.md#l2-output-commitment-construction).
+
+1. A `WithdrawalMessageInclusionProof` is an MPT proof encoded in a struct as follows:
+
+   ```js
+   struct WithdrawalMessageInclusionProof {
+       WithdrawalMessage message;
+       bytes32 l2withdrawalsRoot;
+       bytes memory _key;      // storage key of the withdrawal message commitment
+       bytes memory _value;    // Always bytes32(1) (boolean validity status of the message)
+       bytes memory _proof;    // MPT inclusion proof for the key/value
+   }
+   ```
+1. This document also refers to the `L2Output` type as defined in the
+  [L2 output commitment construction](./proposals.md#l2-output-commitment-construction).
+
+## Withdrawal verification
+
+In order to verify a `WithdrawalMessage`, the following inputs must be provided:
+
+| Type                              | Name               |
+| --------------------------------- | ------------------ |
+| `WithdrawalMessage`               | `message`          |
+| `L2OutputTimestamp`               | `timestamp`        |
+| `WithdrawalsRootInclusionProof`   | `storageRootProof` |
+| `WithdrawalMessageInclusionProof` | `messageProof`     |
+
+These inputs must satisfy the following conditions:
+
+1. The `timestamp` is at least `FINALIZATION_WINDOW` seconds old.
+1. `OutputOracle.l2Outputs(timestamp)` returns a non-zero value `l2Output`.
+1. The `storageRootProof` is valid for the `l2Output` returned in step 2, according to the SSZ encoding described in the
+   [L2 output commitment construction](./proposals.md#l2-output-commitment-construction).
+1. The `messageProof` is valid for the provided `message` and `messageProof`
+
+## Withdrawal Flow
+
+**On L2:**
+
+1. An L2 account sends a withdrawal message (and possibly also ETH) to the `Withdrawoooor` predeploy contract.
+   This is a very simple contract that stores a mapping from the hash of the `WithdrawalMessage` as defined above to a
+   boolean value. (`mapping (bytes32 => bool) withdrawalMessages`)
+2. If ETH is being withdrawn, it can eventually be burned by deploying a contract which immediately `SELFDESTRUCT`s.
+   This has the benefit of using an existing EVM mechanism for removing the ETH from the world state, without having to
+   add a diff to L2 Geth.
+
+**On L1:**
+
+1. A [relayer][g-relayer] submits the required inputs to the `DepositFeed` contract. The relayer may or may not be the same entity which
+   initiated the withdrawal on L2.
```

Comment:
Wait, why are we tangling withdrawals with deposits??

Reply:
Do we in general expect users to send both transactions, being both withdrawer and relayers? If not, then a mechanism for including a relayer payment might be in scope here (probably for later though, after the sequencing milestone).

Reply:
This section should state explicitly that relaying must occur at least 7 days (`FINALIZATION_WINDOW` seconds) after the output commitment for the withdrawal was submitted to L1.

Reply:
I think `DepositFeed` here should actually be the [`L1CrossDomainMessenger`](https://github.com/ethereum-optimism/optimism/blob/develop/packages/contracts/contracts/L1/messaging/L1CrossDomainMessenger.sol). And btw it's important we maintain backwards compatibility with the interface defined there as much as possible

Reply:
The reason I've suggested handling withdrawals via the `DepositFeed` is because that's where the ETH is, so we need to go through there to handle withdrawals with value.

The alternative would be to deposit ETH into the `L1CrossDomainMessenger`, but then you need to add an auth check in the `DepositFeed` that gives the `L1CrossDomainMessenger` the sole permission to mint ETH on L2. That also forces all ETH deposit transactions to have the L1 messenger as the caller. 

My [preference](https://github.com/ethereum-optimism/optimistic-specs/discussions/242#discussioncomment-2409083) would be to build in separate backwards compatibility affordances for allowing the messenger and bridge contracts to continue functioning.